### PR TITLE
add retry on retrieval of job for param verification, full abort on b…

### PR DESCRIPTION
**Pull request by gabemontero**
…uild if job not found

Fixes https://github.com/openshift/jenkins-sync-plugin/issues/272

@bparees @coreydaley @hrishin @piyush-garg fyi ptal

### CODE REVIEW

**Comment by gabemontero** [comment]
ahhhhhh ... our aos-cd-jobs switch to use 3.11 is bumping heads with the fact that we removed old pipeline plugin in 4.0:

```
[Pipeline] End of Pipeline
java.lang.NoSuchMethodError: No such DSL method 'openshiftVerifyBuild' found among steps [_OcAction, _OcContextInit, _OcWatch, archive, bat, build, catchError, checkout, container, containerLog, deleteDir, dir, dockerFingerprintFrom, dockerFingerprintRun, echo, envVarsForTool, error, fileExists, findFiles, getContext, git, input, isUnix, jiraComment, jiraIssueSelector, jiraSearch, junit, library, libraryResource, load, lock, mail, milestone, node, nodesByLabel, parallel, podTemplate, powershell, properties, publishHTML, pwd, readFile, readJSON, readManifest, readMavenPom, readProperties, readTrusted, readYaml, resolveScm, retry, script, sh, sha1, sleep, stage, stash, step, svn, tee, timeout, tool, touch, unarchive, unstash, unzip, validateDeclarativePipeline, waitUntil, withContext, withCredentials, withDockerContainer, withDockerRegistry, withDockerServer, withEnv, wrap, writeFile, writeJSON, writeMavenPom, writeYaml, ws, zip] or symbols [all, allOf, always, any, anyOf, apiToken, architecture, archiveArtifacts, artifactManager, authorizationMatrix, batchFile, bitbucket, booleanParam, branch, buildButton, buildDiscarder, buildingTag, caseInsensitive, caseSensitive, certificate, changeRequest, changelog, changeset, checkoutToSubdirectory, choice, choiceParam, clock, cloud, command, configFile, configFileProvider, configMapVolume, containerEnvVar, containerLivenessProbe, containerTemplate, credentials, cron, crumb, default, defaultView, demand, disableConcurrentBuilds, disableResume, docker, dockerCert, dockerfile, downloadSettings, downstream, dumb, durabilityHint, emptyDirVolume, emptyDirWorkspaceVolume, envVar, envVars, environment, equals, expression, file, fileParam, filePath, fingerprint, frameOptions, freeStyle, freeStyleJob, fromScm, fromSource, git, github, githubPush, globalConfigFiles, headRegexFilter, headWildcardFilter, hostPathVolume, hostPathWorkspaceVolume, hyperlink, hyperlinkToModels, inheriting, inheritingGlobal, installSource, isRestartedRun, jdk, jdkInstaller, jgit, jgitapache, jnlp, jobDsl, jobName, kubernetes, label, lastDuration, lastFailure, lastGrantedAuthorities, lastStable, lastSuccess, legacy, legacySCM, list, local, location, logRotator, loggedInUsersCanDoAnything, masterBuild, maven, maven3Mojos, mavenErrors, mavenMojos, mavenWarnings, modernSCM, myView, never, newContainerPerStage, nfsVolume, nfsWorkspaceVolume, node, nodeProperties, nonInheriting, nonStoredPasswordParam, none, not, oc, onFailure, overrideIndexTriggers, paneStatus, parameters, password, pattern, permanent, persistentVolumeClaim, persistentVolumeClaimWorkspaceVolume, pipeline-model, pipelineTriggers, plainText, plugin, podAnnotation, podEnvVar, pollSCM, portMapping, preserveStashes, projectNamingStrategy, proxy, queueItemAuthenticator, quietPeriod, remotingCLI, run, runParam, schedule, scmRetryCount, scriptApprovalLink, search, secretEnvVar, secretVolume, security, shell, skipDefaultCheckout, skipStagesAfterUnstable, slave, sourceRegexFilter, sourceWildcardFilter, sshUserPrivateKey, stackTrace, standard, status, string, stringParam, swapSpace, tag, text, textParam, tmpSpace, toolLocation, unsecured, upstream, usernameColonPassword, usernamePassword, viewsTabBar, weather, zfs, zip] or globals [currentBuild, docker, env, fileLoader, openshift, params, pipeline, scm]
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:201)
```

Either need to update the plugin docker files for PRs to reference the 3.11 image, or add the old plugin in those dockerfiles to make things kosher

**Comment by gabemontero** [comment]
the login and client plugins don't hit this (those were the plugins we tried previously) but the sync plugin is a more expansive suite of tests

**Comment by gabemontero** [comment]
i've updated the PR dockerfile for sync plugin

File: src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java [hunk 1]
```diff
@@ -119,17 +120,42 @@ public static String getRootUrl() {
...
+            WorkflowJob job = null;
+            long now = System.currentTimeMillis();
+            boolean anyErrors = false;
+            while (System.currentTimeMillis() - 5000 > now) {
+                job = Jenkins.getActiveInstance()
+                        .getItemByFullName(fullName,
+                                WorkflowJob.class);
+                
+                if (job != null) {
+                    if (anyErrors)
```
Comment (hrishin) [inline src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java]: Any reason for logging found message only in retry case `if(anyErrors)`

File: src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java [hunk 2]
```diff
@@ -119,17 +120,42 @@ public static String getRootUrl() {
...
         if (paramMap != null) {
             String fullName = workflowJob.getFullName();
-            WorkflowJob job = Jenkins.getActiveInstance()
-                    .getItemByFullName(fullName,
-                            WorkflowJob.class);
-            if (job == null) {
+            WorkflowJob job = null;
+            long now = System.currentTimeMillis();
+            boolean anyErrors = false;
+            while (System.currentTimeMillis() - 5000 > now) {
```
Comment (hrishin) [inline src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java]: refactor: _may improve readbility_
it's better to give semantic meaning for it?
```
while(forSeconds(5)) {
....
}

boolean forSeconds(int seconds) {
  return System.currentTimeMillis() - seconds > now
}
```

File: src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java]: I did not want to flood the log with error messages when things are down.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java]: Sorry, meant add logs in the normal case

**Comment by gabemontero** [comment]
the new error logic is kicking in when I wouldn't expect it ... investigating

**Comment by gabemontero** [comment]
ok I've pushed some debug and a possible fix ... as i'm in the middle of some 4.0 dev struggles can't test this locally so will rely on PR testing for now

**Comment by gabemontero** [comment]
note in addition to the do / while I added some parens in the conditional ... doesn't seem like it should matter, but the lack of the warning messages in the pod logs would imply we never got into the while loop

File: src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java (hunk 1 above)
Comment (hrishin) [reply src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java]: acknowledged @gabemontero

**Comment by gabemontero** [comment]
Ok .... my manually testing is uncovering some unfortunate intermittent results
- sometimes the retrieval of the job from jenkins works
- when it doesn't, I'm seeing situations where my throwing the abort exception is trashing logs I made right before the abort ... if I insert a sleep, I see the logs
- if i wait a minute or two after a failure, and try a build again, it then is able to retrieve the job from jenkins and things proceed

***SIGH***

Exploring options to get notified or ensure the job is accessible.

**Comment by gabemontero** [comment]
To add to the misery .... after aborting a few times once the job appears accessible in Jenkins, if I submitted multiple jobs as a result, they'll run one after other.

**Comment by gabemontero** [comment]
OK I believe I've somehow manage to salvage things ...
- I discovered how the bc -> job create was doing some jenkins magic to force the write to the xml file ... exactly duplicated that for the env injection piece
- also discovered some paths where we needed to acl up

With those, the pretty consistent can't find job scenario I had today disappeared.

Man I *love* jenkins .... (hopefully sarcasm is discernible)

**Comment by gabemontero** [comment]
@hrishin note the changes in the fix ... I also found a problem with jenkinsJobFullName for looking up the job when the bc to job cache was not populated (after a restart for example)

**Comment by gabemontero** [comment]
v1.0.30 of the plugin has been initiated

**Pull request merged**